tests/unit/has.ts

```diff
@@ -135,6 +135,50 @@ registerSuite({
 			}, TypeError, 'exists and overwrite not true');
 		},
 
+		'works with thenable'(this: any) {
+			const dfd = this.async();
+
+			const thenable = {
+				then(resolve: (_: number) => void) {
+					setTimeout(() => {
+						resolve(5);
+					}, 10);
+
+					return thenable;
+				}
+			};
+
+			hasAdd('thenable', thenable);
+			assert.isFalse(has('thenable'));
+
+			setTimeout(dfd.callback(() => {
+				assert.equal(has('thenable'), 5);
+			}), 100);
+		},
+
+		'failed thenable removes itself from cache'(this: any) {
+			const dfd = this.async();
+
+			const thenable = {
+				then(_: (_: number) => void, reject: (_: Error) => void) {
+					setTimeout(() => {
+						reject(new Error('test error'));
+					}, 10);
+
+					return thenable;
+				}
+			};
+
+			hasAdd('thenable', thenable);
+			assert.isFalse(has('thenable'));
+
+			setTimeout(dfd.callback(() => {
```

Comment:
yeah that was a stupid suggestion on my part :doh: still would perhaps be nice to do something like and just return the promise.

```ts
const promise = new Promise((resolve) => setTimeout(resolve, 100));
return promise.then(() => {
    assert.equal(has('thenable'), 5);
});
```

or 

```ts
function delay(delayTime: number): Promise {
     return new Promise((resolve) => setTimeout(resolve, 100));
}

and just use `return delay(100).then(() => {....`
```